BarrysPeanutsTemporalJava/src/main/java/barryspeanuts/ShoppingCartWorkflowImpl.java

```diff
@@ -0,0 +1,98 @@
+package barryspeanuts;
+
+import barryspeanuts.model.Address;
+import barryspeanuts.model.CreditCard;
+import barryspeanuts.model.PurchaseItem;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.WorkflowQueue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+
+public class ShoppingCartWorkflowImpl implements ShoppingCartWorkflow {
+  private static final Logger logger = Workflow.getLogger(ShoppingCartWorkflowImpl.class);
+
+  boolean exit = false;
+  List<PurchaseItem> purchaseItems = new ArrayList<>();
+  private final WorkflowQueue<Runnable> queue = Workflow.newWorkflowQueue(1024);
+
+  @Override
+  public void startWorkflow() {
+    logger.info("Starting Workflow for Barry's Peanuts");
+    Workflow.await(() -> exit);
+  }
+
+  @Override
+  public List<PurchaseItem> queryPurchaseItems() {
+    return this.purchaseItems;
+  }
+
+  @Override
+  public void addItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.add(purchaseItem);
+  }
+
+  @Override
+  public void addItems(List<PurchaseItem> purchaseItems) {
+    this.purchaseItems.addAll(purchaseItems);
+  }
+
+  @Override
+  public void removeItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.remove(purchaseItem);
+  }
+
+  @Override
+  public void checkOut() {
```

Comment:
Sorry I didn't think to ask this before, but why are `checkOut` and `pay` and `ship` separate? Usually you want these transactional and to all succeed together or fail together as one call. Checkout is a single action from the user side, not three.

I'd just have:

```java
@Override
public void checkOut(CheckoutDetails details) {
```

And that details would have billing address, shipping address, and payment info. No need to mutate purchase items. No need for three separate steps and unnecessary complexity.

Reply:
This is for demonstration purposes only @cretz . The goal of the project is to demonstrate how adapt the Actor model to a Temporal Workflow. Hence the use of signals and the separating business processes into separate functions. I trust you understand.

Reply:
I understand your position and of course, I will defer to your expertise. To my thinking, the additional benefit of breaking things up into those separate functions with apparent signals in the WebUI is that it will make is a lot easier to explain event sourcing the article. For example:

_When it comes time to replenish the workflow after system failure, event sourcing under Temporal makes it possible to recreate the workflow by replaying the events accordingly. Thus, during replay, Temporal will replay each event in the workflow process according to the signal events._

But, nonetheless, let me refactor the code and rewrite the article according to your approach @cretz .

Reply:
I think that makes sense for add items of course. Just not sure breaking up "place order" into three steps makes sense.

Reply:
Refactored by putting `pay` and `ship` as constituent to `checkout`. Addressed.

Reply:
addressed